pkg/ingest/aggregator.go

```diff
@@ -0,0 +1,320 @@
+			return as, errors.Errorf("Incoming chunks are not sorted by timestamp: expected %s after %s", t, as.maxTime)
+		}
+		as.maxTime = t
+		as.count += 1
+		as.sum += v
+		if as.max < v {
+			as.max = v
+		}
+		if as.min > v {
+			as.min = v
+		}
+		if !i.Next() {
+			break
+		}
+	}
+	return as, nil
+}
+
+func (a *aggregator) addSeriesToDf(as *aggregatedSeries) {
+	rs, ok := a.df.seriesRecordSets[as.hash]
+	if !ok {
+		rs = a.df.addRecordSet(as.labels)
+	}
+
+	vals := map[string]interface{}{
+		"_sample_start": as.sampleStart,
+		"_sample_end":   as.sampleEnd,
+		"_min_time":     as.minTime,
+		"_max_time":     as.maxTime,
+	}
+
+	for _, l := range as.labels {
+		if l.Name == "__name__" {
+			continue
+		}
+		vals[l.Name] = l.Value
+	}
+
+	if a.aggrsOptions.Count.Enabled {
+		vals[a.aggrsOptions.Count.Column] = as.count
+	}
+	if a.aggrsOptions.Sum.Enabled {
+		vals[a.aggrsOptions.Sum.Column] = as.sum
+	}
+	if a.aggrsOptions.Min.Enabled {
+		vals[a.aggrsOptions.Min.Column] = as.min
+	}
+	if a.aggrsOptions.Max.Enabled {
+		vals[a.aggrsOptions.Max.Column] = as.max
+	}
+	rs.Records = append(rs.Records, Record{Values: vals})
+}
+
+// Add the active aggregated series into the final dataframe when we've reached the
+// sample end time. Returns pointer to a new instance of the aggregatedSeries
+func (a *aggregator) finalizeSample(as *aggregatedSeries, nextT time.Time) *aggregatedSeries {
+	if as.count > 0 {
+		a.addSeriesToDf(as)
+	}
+
+	// calculate the next sample cycle to contain the nextT time. First calculate how many
+	// whole resolution cycles are between current sampleStart and nextT and then add
+	// those cycles to the current sampleStart
+	nextSampleCycle := ((nextT.Unix() - as.sampleStart.Unix()) / (int64)(a.Resolution/time.Second))
+	nextSampleStart := as.sampleStart.Add((time.Duration(nextSampleCycle)) * a.Resolution)
+
+	as = &aggregatedSeries{
+		labels:      as.labels,
+		hash:        as.hash,
+		sampleStart: nextSampleStart,
+		sampleEnd:   nextSampleStart.Add(a.Resolution),
+	}
+	a.activeSeries[as.hash] = as
+	return as
+}
+
+// Ingest the data to an aggregated set.
+func (a *aggregator) Ingest(s input.Series) error {
+	if s.MinTime().IsZero() {
+		// Zere means no chunks in the series
+		return nil
+	}
+	ls := s.Labels()
+	seriesHash := ls.Hash()
+
+	as, ok := a.activeSeries[seriesHash]
+	if !ok {
+		minTime := s.MinTime()
+		sampleStart := a.aggrsOptions.initSampleTimeFunc(a.Resolution, minTime)
+		sampleEnd := sampleStart.Add(a.Resolution)
+		as = &aggregatedSeries{labels: ls, hash: seriesHash, sampleStart: sampleStart, sampleEnd: sampleEnd}
+		a.activeSeries[seriesHash] = as
+	}
+
+	i, err := s.ChunkIterator()
+	if err != nil {
+		return errors.Wrap(err, "Error while decoding a chunk")
+	}
+
+	if !i.Seek(timestamp.FromTime(as.sampleStart)) {
+		// no chunks after the sampleStart to process
+		return nil
+	}
+
+	as, err = a.ingestChunk(as, i)
+	if err != nil {
+		return errors.Wrap(err, "Error while ingesting a chunk")
```

Comment:
ideally we can keep error lower case. It matter when all of them are wrapped, so 
```suggestion
		return errors.Wrap(err, "error while ingesting a chunk")
```